Drops all 'ifAlias' labels that are just empty strings.

nkinkade opened the pull request:
In other words, only keep the label if it has a value. The vast majority of `ifAlias` labels will have no value. Don't pollute the datastore with empty labels, but instead just drop empty ones and only keep ones with some value.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/m-lab/prometheus-support/107)
<!-- Reviewable:end -->

stephen-soltesz commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[config/federation/prometheus/prometheus.yml, line 411 at r1](https://reviewable.io:443/reviews/m-lab/prometheus-support/107#-KxFOzc7stRcb6y7oXNY:-KxFO3FgrVcBDxmz7z5B:b-gpeb07) ([raw file](https://github.com/m-lab/prometheus-support/blob/84d0388608edb78f97ca51b59d980f17cd5984f5/config/federation/prometheus/prometheus.yml#L411)):*
> ```YAML
>         replacement: snmp-exporter.${1}.measurementlab.net:9116
> 
>       # Most all 'ifAlias' labels will be empty. Rather than pollute the
> ```

Hmm. I've never used `labelkeep` before.

First thing, is I think we'll want this logic on a metric_relabel_configs section.  Since the ifAlias is defined on data scraped by prometheus.

Second thing, I'm not sure now if prometheus does what we want. The docstring for keeplabel says "Match regex against all label names. Any label that does not match will be removed from the set of labels."  That sounds like it's looking at the label name "ifAlias" rather than the value.

We can try this of course, the docs could be misleading.

But, if it does mean what it looks like, then it's not apparent which of the other options will do what we want either.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/prometheus-support/107)*
<!-- Sent from Reviewable.io -->

nkinkade commented:
PTAL?

---

Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[config/federation/prometheus/prometheus.yml, line 411 at r1](https://reviewable.io:443/reviews/m-lab/prometheus-support/107#-KxFOzc7stRcb6y7oXNY:-KxFRY9setSf7irXWrPn:bbo2ba0) ([raw file](https://github.com/m-lab/prometheus-support/blob/84d0388608edb78f97ca51b59d980f17cd5984f5/config/federation/prometheus/prometheus.yml#L411)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

Hmm. I've never used `labelkeep` before.

First thing, is I think we'll want this logic on a metric_relabel_configs section.  Since the ifAlias is defined on data scraped by prometheus.

Second thing, I'm not sure now if prometheus does what we want. The docstring for keeplabel says "Match regex against all label names. Any label that does not match will be removed from the set of labels."  That sounds like it's looking at the label name "ifAlias" rather than the value.

We can try this of course, the docs could be misleading.

But, if it does mean what it looks like, then it's not apparent which of the other options will do what we want either.
</blockquote></details>

Thanks for catching my error on using `relabel_configs` instead of `metric_relabel_configs`. Fixed, I think.

The documentation you point out gave me pause as well. But after going over it a few times I concluded that it simply meant it would iterate over the 'label names' specified in the `source_labels` declaration, in this case being only `[ifAlias]`.  However, since you called it out, I dug into it more and found that your interpretation was correct:

https://github.com/prometheus/prometheus/issues/2367
https://github.com/prometheus/prometheus/issues/2368

Indeed, apparently `source_labels` and `labelkeep` are mutually exclusive.

The links above, plus another one I found seem to indicate you can remove a label by setting `replacement: ""`. I've modified the rules.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/prometheus-support/107#-:-KxFWBAhMCvXiAtMRW6m:b-833305)*
<!-- Sent from Reviewable.io -->

stephen-soltesz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[config/federation/prometheus/prometheus.yml, line 411 at r1](https://reviewable.io:443/reviews/m-lab/prometheus-support/107#-KxFOzc7stRcb6y7oXNY:-KxF_nN-rDtxOmMTtVAu:bjckrw2) ([raw file](https://github.com/m-lab/prometheus-support/blob/84d0388608edb78f97ca51b59d980f17cd5984f5/config/federation/prometheus/prometheus.yml#L411)):*
<details><summary><i>Previously, nkinkade wrote…</i></summary><blockquote>

Thanks for catching my error on using `relabel_configs` instead of `metric_relabel_configs`. Fixed, I think.

The documentation you point out gave me pause as well. But after going over it a few times I concluded that it simply meant it would iterate over the 'label names' specified in the `source_labels` declaration, in this case being only `[ifAlias]`.  However, since you called it out, I dug into it more and found that your interpretation was correct:

https://github.com/prometheus/prometheus/issues/2367
https://github.com/prometheus/prometheus/issues/2368

Indeed, apparently `source_labels` and `labelkeep` are mutually exclusive.

The links above, plus another one I found seem to indicate you can remove a label by setting `replacement: ""`. I've modified the rules.
</blockquote></details>

Okay. Interesting. Thank you for doing the research. It's funny b/c on it's face that looks like a no-op -- replace the empty value with an empty value, but if that does the trick, that's great.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/prometheus-support/107#-:-KxFaNRwxuXKmc099D-C:bnfp4nl)*
<!-- Sent from Reviewable.io -->

nkinkade commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[config/federation/prometheus/prometheus.yml, line 411 at r1](https://reviewable.io:443/reviews/m-lab/prometheus-support/107#-KxFOzc7stRcb6y7oXNY:-KxFbt9o2dnIAKwgGRvD:b6qmw6q) ([raw file](https://github.com/m-lab/prometheus-support/blob/84d0388608edb78f97ca51b59d980f17cd5984f5/config/federation/prometheus/prometheus.yml#L411)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

Okay. Interesting. Thank you for doing the research. It's funny b/c on it's face that looks like a no-op -- replace the empty value with an empty value, but if that does the trick, that's great.
</blockquote></details>

I don't know if that will do the trick, actually. I'm just going off of these:

https://groups.google.com/forum/#!msg/prometheus-developers/zLf13CKY8uY/aXAyvW7bBwAJ
https://github.com/prometheus/prometheus/issues/2367#issuecomment-275159798

These comments would make it appear that setting `replacement: ""` is some sort of undocumented feature that will actually just remove the entire label, and not actually just set the value to an empty string, like intuition would tell me. We'll find out soon, though. :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/prometheus-support/107)*
<!-- Sent from Reviewable.io -->
